File: api/examples/bot.py
"""
Amoebas — Python example bot
Requires: Python 3.8+, requests library  (pip install requests)

Strategy:
  1. Flee any amoeba with a higher score that's within danger range
  2. Otherwise, move toward the nearest pellet
  3. Auto-respawn on death

Usage:
  python bot.py                          # connects to localhost:3001
  python bot.py http://your-server:3001  # custom server
"""

import sys
import time
import math
import requests

BASE_URL = sys.argv[1].rstrip('/') if len(sys.argv) > 1 else 'http://localhost:3001'

DANGER_RADIUS_PADDING = 80   # extra px beyond amoeba radii to start fleeing
TICK_SLEEP = 1 / 30          # aim for ~30 Hz (server also ticks at 30 Hz)


def mass_from_score(score):
    return 8 + 0.05 * score * score


def radius_from_score(score):
    return math.sqrt(mass_from_score(max(0, min(100, score)))) * 2.8


def dist(a, b):
    dx = a['x'] - b['x']
    dy = a['y'] - b['y']
    return math.sqrt(dx * dx + dy * dy)


def normalize(dx, dy):
    mag = math.sqrt(dx * dx + dy * dy)
    if mag < 0.001:
        return 0.0, 0.0
    return dx / mag, dy / mag


def join(name, color):
    r = requests.post(f'{BASE_URL}/api/join', json={'name': name, 'color': color}, timeout=5)
    r.raise_for_status()
    return r.json()


def get_state(token):
    r = requests.get(
        f'{BASE_URL}/api/state',
        headers={'Authorization': f'Bearer {token}'},
        timeout=5
    )
    r.raise_for_status()
    return r.json()


def send_move(token, x, y):
    requests.post(
        f'{BASE_URL}/api/move',
        json={'x': x, 'y': y},
        headers={'Authorization': f'Bearer {token}'},
        timeout=5
    )


def respawn(token):
    r = requests.post(
        f'{BASE_URL}/api/respawn',
        headers={'Authorization': f'Bearer {token}'},
        timeout=5
    )
    return r.json()


def leave(token):
    requests.delete(
        f'{BASE_URL}/api/leave',
        headers={'Authorization': f'Bearer {token}'},
        timeout=5
    )


def decide_direction(state):
    """
    Returns (x, y) move vector, or None to hold still.
    Replace this function with your own logic — LLM, RL model, etc.
    """
    me = state['your_player']
    players = state['players']
    pellets = state['pellets']

    my_radius = radius_from_score(me['score'])

    # --- Threat check: flee any bigger amoeba within danger range ---
    flee_x, flee_y = 0.0, 0.0
    for p in players:
        if p['id'] == me['id']:
            continue
        if p['score'] <= me['score']:
            continue

        their_radius = radius_from_score(p['score'])
        danger_dist = my_radius + their_radius + DANGER_RADIUS_PADDING
        d = dist(me, p)

        if d < danger_dist:
            # Flee: vector pointing away from threat, weighted by proximity
            weight = (danger_dist - d) / danger_dist
            flee_x += (me['x'] - p['x']) * weight
            flee_y += (me['y'] - p['y']) * weight

    if math.sqrt(flee_x ** 2 + flee_y ** 2) > 0.01:
        return normalize(flee_x, flee_y)

    # --- No threat: seek nearest pellet ---
    if not pellets:
        return 0.0, 0.0

    nearest = min(pellets, key=lambda p: (p['x'] - me['x']) ** 2 + (p['y'] - me['y']) ** 2)
    dx = nearest['x'] - me['x']
    dy = nearest['y'] - me['y']
    return normalize(dx, dy)


def main():
    print(f'Connecting to {BASE_URL}')

    # Fetch game info so the bot (or an LLM driver) can read the rules
    info = requests.get(f'{BASE_URL}/api/info', timeout=5).json()
    print(f"World: {info['world']['width']}x{info['world']['height']}")
    print(f"Tips: {info['strategy_tips'][0]}")

    data = join('PyBot', '#4CC9F0')
    token = data['token']
    print(f"Joined as '{data['name']}' — id: {data['agent_id']}")
    print(f"Spawn: ({data['spawn']['x']:.0f}, {data['spawn']['y']:.0f})")
    print('Press Ctrl+C to quit\n')

    try:
        while True:
            state = get_state(token)
            me = state['your_player']

            if not me['alive']:
                info = me.get('death_info') or {}
                killer = info.get('by') or 'unknown'
                print(f"  Died (absorbed by {killer}, score was {info.get('score', '?')}). Respawning…")
                respawn(token)
            else:
                direction = decide_direction(state)
                if direction:
                    send_move(token, direction[0], direction[1])

                # Print status every ~second (every 30 ticks)
                tick = state.get('tick', 0)
                if tick % 30 == 0:
                    alive = len(state['players'])
                    print(f"  tick={tick:6d}  score={me['score']:5.1f}  pos=({me['x']:6.0f},{me['y']:6.0f})  agents_alive={alive}")

            time.sleep(TICK_SLEEP)

    except KeyboardInterrupt:
        print('\nLeaving game…')
        leave(token)
        print('Done.')


if __name__ == '__main__':
    main()

File: api/examples/bot.js
/**
 * Amoebas — Node.js example bot
 * Requires: Node 18+ (built-in fetch)
 *
 * Strategy:
 *   1. Flee any amoeba with a higher score that's within danger range
 *   2. Otherwise, move toward the nearest pellet
 *   3. Auto-respawn on death
 *
 * Usage:
 *   node bot.js                          # connects to localhost:3001
 *   node bot.js http://your-server:3001  # custom server
 */

const BASE_URL = (process.argv[2] || 'http://localhost:3001').replace(/\/$/, '');

const DANGER_RADIUS_PADDING = 80; // extra px beyond amoeba radii to start fleeing
const TICK_MS = 1000 / 30;        // ~30 Hz — same as server tick rate

// ── Physics helpers ────────────────────────────────────────────────────────

function massFromScore(score) {
  return 8 + 0.05 * score * score;
}

function radiusFromScore(score) {
  return Math.sqrt(massFromScore(Math.max(0, Math.min(100, score)))) * 2.8;
}

function dist(a, b) {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  return Math.sqrt(dx * dx + dy * dy);
}

function normalize(dx, dy) {
  const mag = Math.sqrt(dx * dx + dy * dy);
  if (mag < 0.001) return { x: 0, y: 0 };
  return { x: dx / mag, y: dy / mag };
}

// ── API calls ──────────────────────────────────────────────────────────────

async function apiGet(path, token) {
  const headers = token ? { Authorization: `Bearer ${token}` } : {};
  const res = await fetch(`${BASE_URL}${path}`, { headers });
  if (!res.ok) throw new Error(`GET ${path} → ${res.status}: ${await res.text()}`);
  return res.json();
}

async function apiPost(path, body, token) {
  const headers = { 'Content-Type': 'application/json' };
  if (token) headers.Authorization = `Bearer ${token}`;
  const res = await fetch(`${BASE_URL}${path}`, {
    method: 'POST',
    headers,
    body: JSON.stringify(body)
  });
  if (!res.ok && res.status !== 409) throw new Error(`POST ${path} → ${res.status}: ${await res.text()}`);
  return res.json();
}

async function apiDelete(path, token) {
  const res = await fetch(`${BASE_URL}${path}`, {
    method: 'DELETE',
    headers: { Authorization: `Bearer ${token}` }
  });
  return res.json();
}

// ── Decision logic ─────────────────────────────────────────────────────────

/**
 * Returns { x, y } move vector or null to hold still.
 * Replace with your own logic — LLM call, RL model, etc.
 */
function decideDirection(state) {
  const me = state.your_player;
  const { players, pellets } = state;
  const myRadius = radiusFromScore(me.score);

  // Threat check: flee bigger amoebas within danger range
  let fleeX = 0;
  let fleeY = 0;

  for (const p of players) {
    if (p.id === me.id || p.score <= me.score) continue;

    const theirRadius = radiusFromScore(p.score);
    const dangerDist = myRadius + theirRadius + DANGER_RADIUS_PADDING;
    const d = dist(me, p);

    if (d < dangerDist) {
      const weight = (dangerDist - d) / dangerDist;
      fleeX += (me.x - p.x) * weight;
      fleeY += (me.y - p.y) * weight;
    }
  }

  if (Math.sqrt(fleeX ** 2 + fleeY ** 2) > 0.01) {
    return normalize(fleeX, fleeY);
  }

  // No threat: seek nearest pellet
  if (!pellets.length) return { x: 0, y: 0 };

  const nearest = pellets.reduce((best, p) => {
    const d = (p.x - me.x) ** 2 + (p.y - me.y) ** 2;
    return d < best.d ? { p, d } : best;
  }, { p: pellets[0], d: Infinity }).p;

  return normalize(nearest.x - me.x, nearest.y - me.y);
}

// ── Main loop ──────────────────────────────────────────────────────────────

async function main() {
  console.log(`Connecting to ${BASE_URL}`);

  // Fetch game info — useful if you want to feed rules to an LLM before playing
  const info = await apiGet('/api/info');
  console.log(`World: ${info.world.width}x${info.world.height}`);
  console.log(`Tip: ${info.strategy_tips[0]}\n`);

  const joined = await apiPost('/api/join', { name: 'JSBot', color: '#06D6A0' });
  const token = joined.token;
  console.log(`Joined as '${joined.name}' — id: ${joined.agent_id}`);
  console.log(`Spawn: (${joined.spawn.x.toFixed(0)}, ${joined.spawn.y.toFixed(0)})`);
  console.log('Press Ctrl+C to quit\n');

  let running = true;
  process.on('SIGINT', () => { running = false; });

  while (running) {
    const start = Date.now();

    try {
      const state = await apiGet('/api/state', token);
      const me = state.your_player;

      if (!me.alive) {
        const info = me.death_info || {};
        console.log(`  Died (absorbed by ${info.by || 'unknown'}, score was ${info.score ?? '?'}). Respawning…`);
        await apiPost('/api/respawn', {}, token);
      } else {
        const dir = decideDirection(state);
        if (dir) await apiPost('/api/move', dir, token);

        if (state.tick % 30 === 0) {
          const alive = state.players.length;
          console.log(`  tick=${String(state.tick).padStart(6)}  score=${me.score.toFixed(1).padStart(5)}  pos=(${me.x.toFixed(0).padStart(6)},${me.y.toFixed(0).padStart(6)})  agents_alive=${alive}`);
        }
      }
    } catch (err) {
      console.error(`  Error: ${err.message}`);
    }

    const elapsed = Date.now() - start;
    const wait = Math.max(0, TICK_MS - elapsed);
    await new Promise((r) => setTimeout(r, wait));
  }

  console.log('\nLeaving game…');
  await apiDelete('/api/leave', token);
  console.log('Done.');
}

main().catch((err) => { console.error(err); process.exit(1); });
